gitlab/resource_gitlab_project_mirror.go

```diff
@@ -0,0 +1,170 @@
+
+	options := &gitlab.AddProjectMirrorOptions{
+		URL:                   &URL,
+		Enabled:               &enabled,
+		OnlyProtectedBranches: &onlyProtectedBranches,
+		KeepDivergentRefs:     &keepDivergentRefs,
+	}
+
+	log.Printf("[DEBUG] create gitlab project mirror for project %v", projectID)
+
+	mirror, _, err := client.ProjectMirrors.AddProjectMirror(projectID, options)
+	if err != nil {
+		return err
+	}
+	d.Set("mirror_id", mirror.ID)
+
+	mirrorID := strconv.Itoa(mirror.ID)
+	d.SetId(buildTwoPartID(&projectID, &mirrorID))
+	return resourceGitlabProjectMirrorRead(d, meta)
+}
+
+func resourceGitlabMirrorUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gitlab.Client)
+
+	mirrorID := d.Get("mirror_id").(int)
+	projectID := d.Get("project_id").(string)
+	enabled := d.Get("enabled").(bool)
+	onlyProtectedBranches := d.Get("only_protected_branches").(bool)
+	keepDivergentRefs := d.Get("keep_divergent_refs").(bool)
+
+	options := gitlab.EditProjectMirrorOptions{
+		Enabled:               &enabled,
+		OnlyProtectedBranches: &onlyProtectedBranches,
+		KeepDivergentRefs:     &keepDivergentRefs,
+	}
+	log.Printf("[DEBUG] update gitlab project mirror %v for %s", mirrorID, projectID)
+
+	_, _, err := client.ProjectMirrors.EditProjectMirror(projectID, mirrorID, &options)
+	if err != nil {
+		return err
+	}
+	return resourceGitlabProjectMirrorRead(d, meta)
+}
+
+// Documented remote mirrors API does not support a delete method, instead mirror is disabled.
+func resourceGitlabMirrorDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gitlab.Client)
+
+	enabled := false
+
+	mirrorID := d.Get("mirror_id").(int)
+	projectID := d.Get("project_id").(string)
+	onlyProtectedBranches := d.Get("only_protected_branches").(bool)
+	keepDivergentRefs := d.Get("keep_divergent_refs").(bool)
+
+	options := gitlab.EditProjectMirrorOptions{
+		Enabled:               &enabled,
+		OnlyProtectedBranches: &onlyProtectedBranches,
+		KeepDivergentRefs:     &keepDivergentRefs,
+	}
+	log.Printf("[DEBUG] Disable gitlab project mirror %v for %s", mirrorID, projectID)
+
+	_, _, err := client.ProjectMirrors.EditProjectMirror(projectID, mirrorID, &options)
+
+	return err
+}
+
+func resourceGitlabProjectMirrorRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gitlab.Client)
+	mirrorID := d.Get("mirror_id").(int)
+	projectID := d.Get("project_id").(string)
+	log.Printf("[DEBUG] read gitlab project mirror %s id %v", projectID, mirrorID)
+
+	mirrors, _, err := client.ProjectMirrors.ListProjectMirror(projectID)
+
+	if err != nil {
+		return err
+	}
+
+	var mirror *gitlab.ProjectMirror
+	found := false
+
+	for _, m := range mirrors {
+		log.Printf("[DEBUG] project mirror found %v", m.ID)
+		if m.ID == mirrorID {
+			mirror = m
+			found = true
+			return nil
+		}
+	}
+
+	if !found {
+		return fmt.Errorf("unable to find mirror %v on project %s", mirrorID, projectID)
```

Comment:
Instead of returning an error, can you `setId("")`? That seems to be the best practice for handling missing upstream resources.